flink-runtime-web/web-dashboard/src/app/pages/job/checkpoints/job-checkpoints.component.ts

```diff
@@ -16,45 +16,90 @@
-import { distinctUntilChanged } from 'rxjs/operators';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, OnDestroy, OnInit } from '@angular/core';
+import { forkJoin, of, Subject } from 'rxjs';
+import { catchError, distinctUntilChanged, switchMap, takeUntil } from 'rxjs/operators';
 
-import { CheckpointConfig, CheckpointHistory, Checkpoint, JobDetailCorrect } from 'interfaces';
-import { JobService } from 'services';
+import { CheckpointConfig, CheckpointHistory, Checkpoint, JobDetailCorrect } from '@flink-runtime-web/interfaces';
+import { JobService } from '@flink-runtime-web/services';
+
+import { JobLocalService } from '../job-local.service';
 
 @Component({
   selector: 'flink-job-checkpoints',
   templateUrl: './job-checkpoints.component.html',
   styleUrls: ['./job-checkpoints.component.less'],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class JobCheckpointsComponent implements OnInit {
+export class JobCheckpointsComponent implements OnInit, OnDestroy {
   public readonly trackById = (_: number, node: CheckpointHistory): number => node.id;
 
-  public checkPointStats: Checkpoint;
-  public checkPointConfig: CheckpointConfig;
+  public checkPointStats?: Checkpoint;
+  public checkPointConfig?: CheckpointConfig;
   public jobDetail: JobDetailCorrect;
+  public empty = true;
 
   public moreDetailsPanel = { active: false, disabled: false };
 
-  constructor(private readonly jobService: JobService, private readonly cdr: ChangeDetectorRef) {}
+  private refresh$ = new Subject<void>();
+  private destroy$ = new Subject<void>();
+
+  constructor(
+    private readonly jobService: JobService,
+    private readonly jobLocalService: JobLocalService,
+    private readonly cdr: ChangeDetectorRef
+  ) {}
 
   public ngOnInit(): void {
-    this.jobService.jobDetail$.pipe(distinctUntilChanged((pre, next) => pre.jid === next.jid)).subscribe(data => {
-      this.jobDetail = data;
-      this.jobService.loadCheckpointStats(this.jobDetail.jid).subscribe(stats => {
-        this.checkPointStats = stats;
-        this.cdr.markForCheck();
+    this.refresh$
+      .pipe(
+        switchMap(() =>
+          forkJoin([
+            this.jobService.loadCheckpointStats(this.jobDetail.jid),
+            this.jobService.loadCheckpointConfig(this.jobDetail.jid).pipe(
+              catchError(() => {
+                // cp config error won't block displaying cp stats
+                return of(undefined);
+              })
+            )
+          ])
+        ),
+        takeUntil(this.destroy$)
+      )
+      .subscribe({
+        next: ([stats, config]) => {
+          this.empty = false;
+          this.checkPointStats = stats;
+          this.checkPointConfig = config;
+          this.cdr.markForCheck();
+        },
+        error: () => {
+          // only happens when loadCheckpointStats request fails
+          this.empty = true;
+          this.cdr.markForCheck();
+        }
```

Comment:
I think this should be handled as a `catchError` on the request itself, like for the checkpoint config. The problem with handling the error here is that once an error occurs, `refresh$` will never work again.

Reply:
I get what u mean now. Fixed it already.